tools/wpt/browser.py

```diff
@@ -939,44 +939,180 @@ class Chrome(ChromeChromiumBase):
 
     @property
-    def _chromedriver_api_platform_string(self):
-        """chromedriver.storage.googleapis.com has a different filename for M1 binary,
-        while the snapshot URL has a different directory but the same filename."""
-        if self.platform == "Mac" and uname.machine == "arm64":
-            return "mac_arm64"
-        return self._chromedriver_platform_string
+    def _platform_string_chrome(self):
+        """A string that represents the platform-based suffix
+        of the Chrome for Testing downloads.
+        """
+        if self.platform in ("Linux", "Win"):
+            bits = "64" if uname.machine == "x86_64" else "32"
+        elif self.platform == "Mac":
+            bits = "-arm64" if uname.machine == "arm64" else "-x64"
+        else:
+            bits = ""
+        return f"{self.platform.lower()}{bits}"
 
-    def _get_webdriver_url(self, version, revision=None):
-        """Get a ChromeDriver API URL to download a version of ChromeDriver that matches
-        the browser binary version. Version selection is described here:
-        https://chromedriver.chromium.org/downloads/version-selection"""
-        filename = f"chromedriver_{self._chromedriver_api_platform_string}.zip"
+    @property
+    def _package_name_chrome(self):
+        return f"chrome-{self._platform_string_chrome}"
 
-        version = self._remove_version_suffix(version)
+    def _get_webdriver_url(self, version, channel):
+        """Get a ChromeDriver URL to download a version of ChromeDriver that matches
+        the browser binary version.
+        """
+
+        # Remove version suffix if it exists.
+        if version:
+            version = self._remove_version_suffix(version)
+
+        # Check if a file exists containing the matching ChromeDriver version download URL.
+        download_url_reference_file = os.path.join(
+            self._get_browser_binary_dir(None, channel), "matching_chromedriver_url.txt")
+        if os.path.isfile(download_url_reference_file):
+            self.logger.info("Download info for matching ChromeDriver version found.")
+            with open(download_url_reference_file, "r") as f:
+                return f.read()
+
+        # If no ChromeDriver download URL reference file exists,
+        # try to find a download URL based on the version.
+        self.logger.info(f"Searching for ChromeDriver downloads for version {version}.")
+        _, webdriver_url = self._get_download_urls_by_version(version)
+        if webdriver_url is not None:
+            return webdriver_url
+        # We currently use the latest Chromium revision to get a compatible Chromedriver
+        # the debian installation of Chrome Dev, since it is not available from
+        # the ChromeDriver API. If we've gotten to this point,
+        # it is assumed that this is Chrome Dev used for the CI.
+        # TODO(DanielRyanSmith): Remove this final logic when WPT is changed to use
+        # Chrome for Testing in automated runs.
```

Comment:
Could the `latest-patch-versions-per-build` or `latest-versions-per-milestone` JSON API endpoints be useful here? https://github.com/GoogleChromeLabs/chrome-for-testing#json-api-endpoints

Reply:
Good catch. Yes, this might actually be quite useful 🙂 I'll try incorporating it.

Reply:
I've implemented this in a method `_get_webdriver_url_by_build()`. It should work exactly as we need it to, as long as there will always be a ChromeDriver download URL available that matches versions of Chrome Dev on Linux from now on (This is because the CI and automated runs still use a debian installation of Chrome Dev). Is it safe to assume that we will be able to find ChromeDriver downloads to match the MAJOR.MINOR.BUILD versions of Chrome Dev from here on out?

Reply:
> Is it safe to assume that we will be able to find ChromeDriver downloads to match the MAJOR.MINOR.BUILD versions of Chrome Dev from here on out?

It isn’t: https://github.com/web-platform-tests/wpt/issues/42545 We need to fall back to the `latest-versions-per-milestone` endpoint in case `latest-patch-versions-per-build` doesn’t contain an entry for the `MAJOR.MINOR.BUILD` version.

(But ideally, we would avoid this problem altogether by downloading our `chrome` binaries via CfT as well.)